### PR TITLE
Document proposed v4 serialization format

### LINKED ISSUE #503: Document Treelite serialization format.
Document both the current format (3.x) as well as the 4.0 version.

### PR DESCRIPTION
**Pull request by hcho3**
Closes #503

Rendered doc: https://treelite--505.org.readthedocs.build/en/505/serialization/v4.html

TODOs
- [ ] Study scikit-learn's RandomForest / GradientBoosting and ensure that multi-target models can be represented with the v4 spec.
- [ ] Study XGBoost's multi-target models (both model-per-target and [vector-leaf trees](https://xgboost.readthedocs.io/en/latest/tutorials/multioutput.html)) and ensure that they can be represented with the v4 spec.

### CODE REVIEW

File: docs/serialization/v4.rst [hunk 1]
```diff
@@ -0,0 +1,161 @@
...
+     - ``num_class``: an array of ``uint32_t`` with length ``num_target``.
+     - ``leaf_vector_shape``: an array of ``uint32_t`` with length 2. The first dimension is always ``num_target``. The second dimension is either 1 or ``max(num_class)``.
+     - ``grove_per_class``: single ``bool`` scalar
+
+   * Model parameters (``ModelParam``) a structure with following fields
+
+     - ``pred_transform``: 256-character long ``char`` array. Stores a human-readable name of the transformation function that's applied to prediction outputs. The unused elements in the array should be padded with null characters (``\0``).
+     - ``sigmoid_alpha``: single ``float`` scalar. This model parameter is relevant when ``pred_transform="sigmoid"``.
+     - ``ratio_c``: single ``float`` scalar. This model parameter is relevant when ``pred_transform="exponential_standard_ratio"``.
+   * ``base_scores``: an array of ``float``. This vector is expected to have length ``num_target``.
```
Comment (hcho3) [inline docs/serialization/v4.rst]:
```suggestion
   * ``base_scores``: an array of ``LeafOutputType``. This vector is expected to have length ``num_target``.
```

File: docs/serialization/v4.rst [hunk 2]
```diff
@@ -0,0 +1,161 @@
...
+  - ``kFloat32`` (2)
+  - ``kFloat64`` (3)
+
+* ``TaskType``: underlying type ``uint8_t``. Indicates the type of the learning task.
+
+  - ``kBinaryClf`` (0): binary classifier
+  - ``kRegressor`` (1): regressor
+  - ``kLearningToRank`` (2): learning-to-rank
+  - ``kMultiClfGrovePerClass`` (3): multi-class classifier with grove-per-class
+  - ``kMultiClfProbDistLeaf`` (4): multi-class classifier with vector leaf outputs producing probability distribution
```
Comment (hcho3) [inline docs/serialization/v4.rst:26]:
```suggestion
  - ``kMultiClfProbDistLeaf`` (4): multi-class classifier with vector leaf outputs producing probability distribution
  - ``kIsolationForest`` (5): isolation forest
```

File: docs/serialization/v4.rst [hunk 3]
```diff
@@ -0,0 +1,162 @@
...
+     - ``num_class``: an array of ``uint32_t`` with length ``num_target``.
+     - ``leaf_vector_shape``: an array of ``uint32_t`` with length 2. The first dimension is always ``num_target``. The second dimension is either 1 or ``max(num_class)``.
+     - ``grove_per_class``: single ``bool`` scalar
+
+   * Model parameters (``ModelParam``) a structure with following fields
+
+     - ``pred_transform``: 256-character long ``char`` array. Stores a human-readable name of the transformation function that's applied to prediction outputs. The unused elements in the array should be padded with null characters (``\0``).
+     - ``sigmoid_alpha``: single ``float`` scalar. This model parameter is relevant when ``pred_transform="sigmoid"``.
+     - ``ratio_c``: single ``float`` scalar. This model parameter is relevant when ``pred_transform="exponential_standard_ratio"``.
+   * ``base_scores``: an array of ``LeafOutputType``. This vector is expected to have length ``num_target``.
```
Comment (hcho3) [inline docs/serialization/v4.rst]:
```suggestion
   * ``base_scores``: an array of ``LeafOutputType``. This vector is expected to have length ``num_target * max(num_class)``. The elements will be laid out in the row-major layout.
```

File: docs/serialization/v4.rst [hunk 4]
```diff
@@ -0,0 +1,162 @@
...
+
+     - Number of fields: single ``int32_t`` scalar. Set this value to ``0``, to indicate the lack of optional fields.
+
+#. Tree 1: Use the same set of fields as Tree 0.
+#. Other trees ...
+
+.. note:: Caveat for multi-target, multi-class classifiers
+
+   When the number of classes are different for targets, we use the larget number of
+   classes (``max_num_class``) and use it to shape the leaf vector. The leaf vector
```
Comment (hcho3) [inline docs/serialization/v4.rst]:
```suggestion
   classes (``max_num_class``) and use it to shape the leaf vector (and ``base_scores``). The leaf vector
```

File: docs/serialization/v4.rst [hunk 5]
```diff
@@ -0,0 +1,162 @@
...
+     - ``cright_``: single ``int32_t`` scalar. Indicates the ID of the right child node. Set to ``-1`` to indicate the lack of the right child.
+     - ``sindex_``: single ``uint32_t`` scalar. This field gives both the feature ID used in the current test node (``split_index``), as well as the default direction for the missing value (``default_left``). Set this value by computing ``split_index | (default_left ? (1U << 31U) : 0)``.
+     - ``info_``: a union type containing ``leaf_value`` (of type ``LeafOutputType``) and ``threshold`` (of type ``ThresholdType``). To set this field, determine whether the node is a leaf node or an internal test node. Use ``leaf_value`` for leaf nodes; use ``threshold`` for internal test nodes.
+     - ``comp_``: single ``int8_t`` scalar representing enum ``Operator``.
+     - ``categories_list_right_child_``: single ``bool`` scalar.
+
+   * Leaf vectors
+
+     - Content (``leaf_vector_``): an array of ``LeafOutputType``. This array stores the leaf vectors for all nodes, such that
+       the sub-array ``leaf_vector_[leaf_vector_begin[i]_:leaf_vector_end_[i]]`` yields the leaf vector for the i-th node.
```
Comment (hcho3) [inline docs/serialization/v4.rst:106]:
```suggestion
       the sub-array ``leaf_vector_[leaf_vector_begin[i]_:leaf_vector_end_[i]]`` yields the leaf vector for the i-th node.
       The leaf vector uses the row-major layout to store a 2D array.
```

File: docs/serialization/v4.rst [hunk 6]
```diff
@@ -0,0 +1,163 @@
...
+
+     - Number of fields: single ``int32_t`` scalar. Set this value to ``0``, to indicate the lack of optional fields.
+
+#. Tree 1: Use the same set of fields as Tree 0.
+#. Other trees ...
+
+.. note:: Caveat for multi-target, multi-class classifiers
+
+   When the number of classes are different for targets, we use the larget number of
+   classes (``max_num_class``) and use it to shape the leaf vector (and ``base_scores``). The leaf vector
```
Comment (hcho3) [inline docs/serialization/v4.rst]:
```suggestion
   When the number of classes are different for targets, we use the larget number of
   classes (``max_num_class``) to shape the leaf vector (and ``base_scores``). The leaf vector
```

File: docs/serialization/v4.rst [hunk 7]
```diff
@@ -0,0 +1,163 @@
...
+     - Number of fields: single ``int32_t`` scalar. Set this value to ``0``, to indicate the lack of optional fields.
+
+#. Tree 1: Use the same set of fields as Tree 0.
+#. Other trees ...
+
+.. note:: Caveat for multi-target, multi-class classifiers
+
+   When the number of classes are different for targets, we use the larget number of
+   classes (``max_num_class``) to shape the leaf vector (and ``base_scores``). The leaf vector
+   will have shape ``(num_target, max_num_class)``. This heuristic has the following
```
Comment (hcho3) [inline docs/serialization/v4.rst]:
```suggestion
   will have shape ``(num_target, max_num_class)``, with extra elements padded with ``0``. This heuristic has the following
```

**Review by RAMitchell — APPROVED** [review]
Looks very nice, some clear improvements over v3 in simplifying the TaskTypes and parameters. I can't see anything immediately missing.

**Comment by hcho3** [comment]
@RAMitchell Thanks for your review. I will try to implement a prototype for a few real-world multi-target models and see if the proposed spec is adequate.

**Pull request merged**